cmd/storer2elastic/dataindexer/dataIndexer.go

```diff
@@ -0,0 +1,292 @@
+
+type dataPreparerAndIndexer struct {
+	elasticIndexer    indexer.Indexer
+	databaseReader    DatabaseReaderHandler
+	shardCoordinator  sharding.Coordinator
+	marshalizer       marshal.Marshalizer
+	hasher            hashing.Hasher
+	nodesCoordinators map[uint32]NodesCoordinator
+	hdrsToIndexLater  []data.HeaderHandler
+}
+
+type persistersHolder struct {
+	miniBlocksPersister           storage.Persister
+	transactionPersister          storage.Persister
+	unsignedTransactionsPersister storage.Persister
+	rewardTransactionsPersister   storage.Persister
+}
+
+// Args holds the arguments needed for creating a new dataPreparerAndIndexer
+type Args struct {
+	ElasticIndexer    indexer.Indexer
+	DatabaseReader    DatabaseReaderHandler
+	ShardCoordinator  sharding.Coordinator
+	Marshalizer       marshal.Marshalizer
+	Hasher            hashing.Hasher
+	GenesisNodesSetup sharding.GenesisNodesSetupHandler
+}
+
+// New returns a new instance of dataPreparerAndIndexer
+func New(args Args) (*dataPreparerAndIndexer, error) {
+	if check.IfNil(args.ElasticIndexer) {
+		return nil, ErrNilElasticIndexer
+	}
+	if check.IfNil(args.DatabaseReader) {
+		return nil, ErrNilDatabaseReader
+	}
+	if check.IfNil(args.ShardCoordinator) {
+		return nil, ErrNilShardCoordinator
+	}
+	if check.IfNil(args.Marshalizer) {
+		return nil, ErrNilMarshalizer
+	}
+	if check.IfNil(args.Hasher) {
+		return nil, ErrNilHasher
+	}
+	if check.IfNil(args.GenesisNodesSetup) {
+		return nil, ErrNilGenesisNodesSetup
+	}
+
+	dpi := &dataPreparerAndIndexer{
+		elasticIndexer:   args.ElasticIndexer,
+		databaseReader:   args.DatabaseReader,
+		shardCoordinator: args.ShardCoordinator,
+		marshalizer:      args.Marshalizer,
+		hasher:           args.Hasher,
+		hdrsToIndexLater: make([]data.HeaderHandler, 0),
+	}
+
+	nodesCoordinator, err := dpi.createNodesCoordinators(args.GenesisNodesSetup)
+	if err != nil {
+		return nil, err
+	}
+
+	dpi.nodesCoordinators = nodesCoordinator
+	return dpi, nil
+}
+
+// Index will prepare the data for indexing and will try to index all the data before the given timeout
+func (dpi *dataPreparerAndIndexer) Index(timeout int) error {
+	errChan := make(chan error, 0)
+	go func() {
+		dpi.startIndexing(errChan)
+	}()
+
+	select {
+	case err := <-errChan:
+		return err
+	case <-time.After(time.Duration(timeout) * time.Second):
+		return ErrTimeIsOut
+	}
+}
+
+func (dpi *dataPreparerAndIndexer) startIndexing(errChan chan error) {
+	records, err := dpi.databaseReader.GetDatabaseInfo()
+	if err != nil {
+		errChan <- err
+		return
+	}
+
+	for _, db := range records {
+		hdrs, err := dpi.databaseReader.GetHeaders(db)
+		if err != nil {
+			errChan <- err
+			return
+		}
+
+		persisters, err := dpi.preparePersistersHolder(db)
+		for _, hdr := range hdrs {
+			err = dpi.indexHeader(persisters, db, hdr, db.Epoch)
+			if err != nil {
+				log.Warn("cannot index header", "error", err)
+			}
+		}
+		for _, hdr := range dpi.hdrsToIndexLater {
```

Comment:
because there is a chance that shard headers are read before the meta block for the epoch start corresponding to header's epoch. but this can be fixed easily by indexing meta headers first. I will remove it

Reply:
also. for metablocks. they are currently not read in order so for example for processing the nonce `37` I need the epoch start meta block for it's epoch. but this happens way more few times after changing the sorting algorithm